Dockerfile

```diff
@@ -1,39 +1,76 @@
+ARG devDeps="python39-devel gcc"
 
-RUN microdnf update -y && \
-    microdnf install --setopt=install_weak_deps=0 --setopt=tsflags=nodocs -y \
-    git-core python39 python39-pip tzdata libpq-devel && \
-    rpm -qa | sort > packages-before-devel-install.txt && \
-    microdnf install --setopt=tsflags=nodocs -y python39-devel gcc && \
-    rpm -qa | sort > packages-after-devel-install.txt
+ARG TEST_IMAGE=false
 
-RUN adduser --gid 0 -d /opt/app-root --no-create-home insights
+FROM registry.access.redhat.com/ubi8/ubi-minimal AS build
+
+ARG deps
+ARG devDeps
 
 ENV LC_ALL=C.utf8
 ENV LANG=C.utf8
 
 ENV APP_ROOT=/opt/app-root
-ENV PIPENV_VENV_IN_PROJECT=1
-
-ENV POETRY_CONFIG_DIR=/opt/app-root/.pypoetry/config
-ENV POETRY_DATA_DIR=/opt/app-root/.pypoetry/data
-ENV POETRY_CACHE_DIR=/opt/app-root/.pypoetry/cache
 
+ENV POETRY_NO_INTERACTION=1 \
+    POETRY_VIRTUALENVS_IN_PROJECT=1 \
+    POETRY_VIRTUALENVS_CREATE=1 \
+    POETRY_CONFIG_DIR=/opt/app-root/.pypoetry/config \
+    POETRY_DATA_DIR=/opt/app-root/.pypoetry/data \
+    POETRY_CACHE_DIR=/opt/app-root/.pypoetry/cache
 ENV UNLEASH_CACHE_DIR=/tmp/unleash_cache
 
-COPY . ${APP_ROOT}/src
+RUN microdnf update -y && \
+    microdnf module enable python39 && \
+    microdnf install --setopt=install_weak_deps=0 --setopt=tsflags=nodocs -y $deps && \
+    microdnf install --setopt=tsflags=nodocs -y $devDeps
+RUN pip3 install --force-reinstall poetry~=1.6.0
 
 WORKDIR ${APP_ROOT}/src
 
-RUN pip3 install --upgrade pip && \
-    pip3 install --force-reinstall poetry~=1.6.0
+COPY pyproject.toml poetry.lock ${APP_ROOT}/src
+
+RUN poetry install --sync --no-root && rm -rf $POETRY_CACHE_DIR
+
+#######################
+
+FROM registry.access.redhat.com/ubi8/ubi-minimal AS final
+
+ARG deps
+ARG TEST_IMAGE
+
+ENV LC_ALL=C.utf8
+ENV LANG=C.utf8
+
+ENV APP_ROOT=/opt/app-root
+
+ENV POETRY_NO_INTERACTION=1 \
+    POETRY_VIRTUALENVS_IN_PROJECT=1 \
+    POETRY_VIRTUALENVS_CREATE=1 \
+    POETRY_CONFIG_DIR=/opt/app-root/.pypoetry/config \
+    POETRY_DATA_DIR=/opt/app-root/.pypoetry/data \
+    POETRY_CACHE_DIR=/opt/app-root/.pypoetry/cache
+
+ENV VIRTUAL_ENV_DIR=${APP_ROOT}/src/.venv
+ENV UNLEASH_CACHE_DIR=/tmp/unleash_cache
+
+RUN microdnf update -y && \
+    microdnf module enable python39 && \
+    microdnf install --setopt=install_weak_deps=0 --setopt=tsflags=nodocs -y $deps
+RUN pip3 install --force-reinstall poetry~=1.6.0
 
-RUN chown -R insights:0 /opt/app-root  && \
-    chgrp -R 0 /opt/app-root && \
-    chmod -R g=u /opt/app-root
+RUN adduser --gid 0 -d /opt/app-root --no-create-home insights
 
 USER insights
 
-RUN poetry install --without dev --sync
+WORKDIR ${APP_ROOT}/src
+
+COPY --chown=insights:0 . ${APP_ROOT}/src
+
+COPY --from=build --chown=insights:0 $VIRTUAL_ENV_DIR $VIRTUAL_ENV_DIR
+
+# allows unit tests to run successfully within the container if image is built in "test" environment
+RUN if [ "$TEST_IMAGE" = "true" ]; then chgrp -R 0 $APP_ROOT && chmod -R g=u $APP_ROOT; fi
```

Comment:
Can the tests be updated in the fashion of having them successfully run withought chgrp/chmod mangling in this file?

Reply:
root user should be avoided for applications. `USER 0` is there to install packages, after that there is a flip to a normal user (`USER 1001`).

Reply:
App folder can be set to be under root group to have it readable/writable by a random user (UID) that OpenShift/k8s sets when running containers. (ref: https://developers.redhat.com/articles/2021/11/11/best-practices-building-images-pass-red-hat-container-certification)